### ISSUE TITLE
Empty string for boolean converted to null even though database has default value.

### ISSUE THREAD

**Issue by liubomirwm** [yiisoft/yii2#17106]
### What steps will reproduce the problem?
Have database as that:
![image](https://user-images.githubusercontent.com/16529989/52328173-4dc61980-29f7-11e9-8d57-64c5ede17b26.png)

Leave checkbox empty. Once $model->load() is called in Controller, an empty string is assigned to the model property. On save this exception is given:
![image](https://user-images.githubusercontent.com/16529989/52328278-a5fd1b80-29f7-11e9-86cd-31f6cc889f4b.png)

### What is the expected result?
Save the entry with default value of 0.
### What do you get instead?
Exception:
![image](https://user-images.githubusercontent.com/16529989/52328278-a5fd1b80-29f7-11e9-86cd-31f6cc889f4b.png)

### Additional info

| Q                | A
| ---------------- | ---
| Yii version      | 2.0.16
| PHP version      | 5.6.30
| Operating system | Windows Server 1607 14393.2759

**Comment by rob006** [comment]
Do you have validation rules in you model for converting empty input to `0`?

**Comment by liubomirwm** [comment]
No, i don't have such validation rules in model. I added a check in controller (manually set to false if empty string) and it works now.

What caused me to report was a similar issue #16364.

**Comment by terabytesoftw** [comment]
Closing this as expected behavior rather than a Yii bug.

The column is explicitly included in the INSERT statement with a value coming from the request. Yii typecasts an empty string for non-string DB columns to `null`, and SQL Server does not apply a DEFAULT constraint when a column is explicitly assigned `NULL`; the default is used only when the column is omitted from the INSERT or when the `DEFAULT` keyword is used.

For checkbox inputs, the application should submit `0` for the unchecked state, for example using Yii default `activeCheckbox()` behavior, or normalize the value with a `default`/`filter` validation rule, `loadDefaultValues()`, or `AttributeTypecastBehavior`.

Changing Yii to replace empty strings with database defaults during save would be ambiguous and backward-incompatible.
